### DISCUSSION TITLE
Server Actions Loading State

### DISCUSSION THREAD

**Discussion by rmartin93** [vercel/next.js#58415] [Help]
### Summary

All of the examples on the new Next tutorial show how to use Server Actions to send form data, which is great, but I need a super simple server action that sends no data. I'm literally just trying to refresh some data, and I'd like to have a loading indicator to let the user know the fetch is happening. 

My button
```
"use client";
import { updateUsers } from "@/app/_lib/actions";

export default function RefreshUsersButton() {
	return (
		// Action must be called from a form because in React, the action attribute is considered a special prop - meaning React builds on top of it to allow actions to be invoked.
		<form action={updateUsers}>
			<input type="hidden" name="id" />
			<button type="submit" className="btn btn-primary">
				Refresh Users Data
			</button>
		</form>
	);
}
```

My actions.ts
```
"use server";
import { revalidatePath } from "next/cache";
export async function updateUsers() {
	revalidatePath("/");
}
```

package.json
```
{
	"name": "macro-tracker",
	"version": "0.1.0",
	"private": true,
	"scripts": {
		"dev": "next dev -p 4000",
		"build": "next build",
		"start": "next start",
		"lint": "next lint"
	},
	"dependencies": {
		"@vercel/postgres": "^0.5.1",
		"bootstrap": "^5.3.2",
		"bootstrap-icons": "^1.11.1",
		"next": "14.0.2",
		"react": "^18",
		"react-dom": "^18"
	},
	"devDependencies": {
		"@types/node": "^20",
		"@types/react": "^18",
		"@types/react-dom": "^18",
		"eslint": "^8",
		"eslint-config-next": "14.0.2",
		"typescript": "^5"
	}
}
```

If anybody has an idea on how to get a loading state into the button that would be awesome. Thanks.

**Comment by danieltott** [comment]
Check this out: [Displaying Loading State](https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations#displaying-loading-state)

**Reply by aryanprince** [reply]
They updated the docs recently, so the updated link to the docs with loading/pending state is [here](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#pending-states).

**Comment by aryanprince** [comment]
As @danieltott mentions, best way to display a loading state for your Server Action is by using the `useFormStatus` hook in another child component ([read more here](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#pending-states)).

In your case @rmartin93, you would have to just import the`useFormStatus` hook, create a new `RefreshButton` component (any name is fine), and then use the `pending` state in this new component as you wish. Code example below:

```tsx
"use client";

import { updateUsers } from "@/app/_lib/actions";
import { useFormStatus } from "react-dom"; // Import this

export default function RefreshUsersButton() {
  return (
    <form action={updateUsers}>
      <input type="hidden" name="id" />
      <RefreshButton /> // Use new component here
    </form>
  );
}

// Create this component
function RefreshButton() {
  const { pending } = useFormStatus();

  return (
    <button type="submit" className="btn btn-primary" disabled={pending}>
      Refresh Users Data
    </button>
  );
}
```

Note: As of latest version of Next.js `v14.2.5`, you'll have to create a 'new' component - just to get the loading/pending state. I wish we could directly use the pending state in the same component itself. This is getting changed with the new `useActionState` hook in Next.js/React Canary IIRC.
